File: go/unionfind/union_find3.go
package unionfind

import (
	"errors"
)

// Find2 的优化版本，优化树的合并过程
type Node struct {
	id int
	size int
}

type UnionFind3 struct {
	parent []*Node
}

func NewSizeUnionFind(size int) *UnionFind3 {
	parent := make([]*Node, size)
	for i := 0; i < size; i++ {
		node := &Node{id: i, size: 1}
		parent[i] = node
	}
	return &UnionFind3{parent}
}

func (u *UnionFind3) GetSize() int {
	return len(u.parent)
}

func (u *UnionFind3) find(p int) (node *Node, err error) {
	if p < 0 || p >= len(u.parent) {
		err = errors.New("Index out of range")
		return
	}

	for ; p != u.parent[p].id; p = u.parent[p].id {}
	node = u.parent[p]
	return
}

func (u *UnionFind3) IsConnected(p, q int) bool {
	pPar, err := u.find(p)
	if err != nil {
		return false
	}
	qPar, err := u.find(q)
	if err != nil {
		return false
	}

	return qPar.id == pPar.id
}

func (u *UnionFind3) UnionElements(p, q int) error {
	pPar, err := u.find(p)
	if err != nil {
		return err
	}
	qPar, err := u.find(q)
	if err != nil {
		return err
	}

	if pPar.id == qPar.id {
		return nil
	}

	pSize := pPar.size
	qSize := qPar.size

	if pSize <= qSize {
		pPar.id = qPar.id
		qPar.size = qPar.size + pPar.size
	} else {
		qPar.id = pPar.id
		pPar.size = pPar.size + qPar.size
	}
	return nil
}

File: js2/unionFind3.ts
import type { UnionFind } from './unionFind';

// quick union 版本 基于 size 优化
export default class UnionFind3 implements UnionFind {
    private parent: Array<number>;
    private sizes: Array<number>; 

    constructor(size: number) {
        this.parent = new Array(size);
        this.sizes = new Array(size);

        for (let i = 0; i < size; i++) {
            this.parent[i] = i;
            this.sizes[i] = 1;
        }
    }

    getSize() {
        return this.parent.length;
    }

    private find(p: number): number {
        while (this.parent[p] !== p) {
            p = this.parent[p];
        }
        return p;
    }

    // O(h)
    isConnected(p: number, q: number): boolean {
        return this.find(p) === this.find(q);
    }

    // O(h)
    unionElements(p: number, q: number) {
        const pParent = this.find(p);
        const qParent = this.find(q);
        if (pParent === qParent) {
            return;
        }
        if (this.sizes[pParent] < this.sizes[qParent]) {
            this.parent[pParent] = qParent;
            this.sizes[qParent] += this.sizes[pParent];
        } else {
            this.parent[qParent] = pParent;
            this.sizes[pParent] += this.sizes[qParent];
        }
    }
}